bluebutton-server-app/src/main/java/gov/hhs/cms/bluebutton/server/app/stu3/providers/SamhsaMatcher.java

```
@@ -275,12 +293,64 @@ private boolean isSamhsaIcdDiagnosis(DiagnosisComponent diagnosis) {
 			throw new BadCodeMonkeyException(e);
 		}
 
-		for (Coding diagnosisCoding : diagnosisConcept.getCoding()) {
-			if (IcdCode.CODING_SYSTEM_ICD_9.equals(diagnosisCoding.getSystem())) {
-				if (isSamhsaIcd9Diagnosis(diagnosisCoding))
+		if (diagnosisConcept != null) {
+			for (Coding diagnosisCoding : diagnosisConcept.getCoding()) {
+				if (IcdCode.CODING_SYSTEM_ICD_9.equals(diagnosisCoding.getSystem())) {
+					if (isSamhsaIcd9Diagnosis(diagnosisCoding))
+						return true;
+				} else if (IcdCode.CODING_SYSTEM_ICD_10.equals(diagnosisCoding.getSystem())) {
+					if (isSamhsaIcd10Diagnosis(diagnosisCoding))
+						return true;
+				} else {
+					// Fail safe: if we don't know the ICD version, assume the code is SAMHSA.
 					return true;
-			} else if (IcdCode.CODING_SYSTEM_ICD_10.equals(diagnosisCoding.getSystem())) {
-				if (isSamhsaIcd10Diagnosis(diagnosisCoding))
+				}
+			}
+		}
+
+		CodeableConcept packageConcept = diagnosis.getPackageCode();
+		if (packageConcept != null) {
+			for (Coding packageCoding : packageConcept.getCoding()) {
+				if (SamhsaMatcher.DRG.equals(packageCoding.getSystem())) {
+					if (isSamhsaDrgCode(packageCoding))
+						return true;
+				} else {
+					// Fail safe: if we don't know the package coding system, assume the code is
+					// SAMHSA.
+					return true;
```

Just a note for a future PR: I think we'd be better off throwing `BadCodeMonkeyException`s for things like this, instead of the fail-safe returns. My fault.